addons/suggest_locations/tests/test_stock_move_line.py

```diff
@@ -0,0 +1,237 @@
+        for drop_constraint in ("dont_scan", "scan", "suggest", "enforce"):
+            self.picking_type_pick.u_drop_location_constraint = drop_constraint
+            locs = self.MoveLine.suggest_locations(picking=self.picking, values=values)
+            self.assertEqual(locs, self.test_goodsout_location_01 | self.test_goodsout_location_03)
+        # Add u_drop_location_constraint with empty
+        for drop_constraint in ("suggest_with_empty", "enforce_with_empty"):
+            self.picking_type_pick.u_drop_location_constraint = drop_constraint
+            locs = self.MoveLine.suggest_locations(picking=self.picking, values=values)
+            self.assertEqual(
+                locs,
+                self.test_goodsout_location_01
+                | self.test_goodsout_location_03
+                | self.test_goodsout_location_04,
+            )
+
+    def test06_suggest_locations_limit_results(self):
+        """Get the suggest locations via self and limit the results.
+        This is set up for 'by_product' policy.
+        We loop through all policies in a very simple way
+        """
+        #  Get apple mls
+        apple_mls = self.mls.filtered(lambda ml: ml.product_id == self.apple)
+        # Set u_drop_location_constraint
+        self.picking_type_pick.u_drop_location_constraint = "suggest_with_empty"
+        self.assertEqual(
+            apple_mls.suggest_locations(),
+            self.test_goodsout_location_01
+            | self.test_goodsout_location_03
+            | self.test_goodsout_location_04,
+        )
+        # Limit the results
+        locs = apple_mls.suggest_locations(limit=1)
+        self.assertEqual(len(locs), 1)
+        self.assertIn(
+            locs,
+            self.test_goodsout_location_01
+            or self.test_goodsout_location_03
+            or self.test_goodsout_location_04,
+        )
+
+    def test07_validate_location_dest_nothing_dropppable(self):
+        """Check that we just return when nothing is droppable"""
+        # Complete the moves
+        self.picking.move_lines.quantity_done = 5
+        self.picking.action_done()
+        #  Get apple mls
+        apple_mls = self.mls.filtered(lambda ml: ml.product_id == self.apple)
+        self.assertIsNone(apple_mls.validate_location_dest())
+
+    def test08_validate_location_dest_nothing_dropppable(self):
+        """Check that we just return when the drop location is a view"""
+        # Set locations to a view location
+        self.out_location.write({"usage": "view"})
+        locations = self.out_location
+        self.assertIsNone(self.mls.validate_location_dest(locations))
+
+    def test9_check_validate_on_write(self):
+        """Check the validation policy controls whether a write can be performed
+        for the location destination.
+        """
+        # Set policy to suggest
+        self.picking_type_pick.u_drop_location_constraint = "suggest"
```

Comment:
Can we move enforce first, then change to suggest and repeat with no error, please? Also only using the move line of the product is not in `test_goodsout_location_02`.